core/src/main/scala/org/broadinstitute/dsde/rawls/billing/BpmBillingProjectLifecycle.scala

```diff
@@ -125,7 +125,7 @@ class BpmBillingProjectLifecycle(
           )
         case _ => ()
       }
-      _ <- billingRepository.getLandingZoneId(projectName).map {
+      _ <- billingRepository.getLandingZoneId(projectName).flatMap {
```

Comment:
This code is pretty much identical to the code at line 150 (return type of `Future[Option[String]]`, so I'm guessing this should be a `flatMap`? Or perhaps it doesn't because it's not the last one in the for comprehension?

Still confused….

Reply:
I find it helpful to substitute types into the signatures of `map` and `flatMap`.

Here's the thing we want to use, wrapped up in a `Future`:
```scala
billingRepository.getBillingProfileId(projectName) : Future[Option[String]]`
```

In the case where there's `Some(profileId)`, we want to perform some async computation to fetch all the the other projects that share that profile and delete the profile if none exist. this function has the following signature `Option[String] -> Future[Unit]`. Why? Because `getBillingProjectsWithProfile` is async and returns `Future`.

Let's subsitute this into the sigatures for `map` and `flatMap`, starting with `map`:
```scala
// in the definition of `trait Future[A]`
def map[B](f: A -> B): Future[B]

substitute `A`:
def map[B](f: Option[String] -> B): Future[B]

substitute `f` signature to get `B`:
def map[Future[Unit]](Option[String] -> Future[Unit]): Future[Future[Unit]]
```
And there's our problem - our Futures are interleaved. `Future` is async and runs eagerly in its own thread. If we wait for the outer future to be done, there's no guarantee that the inner future will be done. We need to synchronize/compose these effects so that they're done sequentially. For that, we use `flatMap`: 
```scala
// in the definition of `trait Future[A]`
def flatMap[B](f: A -> Future[B]): Future[B]

substitute `A`:
def flatMap[B](f: Option[String] -> Future[B]): Future[B]

substitute `f` signature to get `B`:
def flatMap[Unit](Option[String] -> Future[Unit]): Future[Unit]
```
Thus, when we wait for the result to be done, we can guarantee that the effects from our continuation, `f`, have been performed.

Reply:
I simplified the code back to the minimum necessary changes. Still interested at some point in discussing the alternative of moving the inner asynchronous call out, just for my own edification.

Reply:
Sure - we can pair on it when you're free if you like.